### PR TITLE
add azure machine pool machine proposal

### PR DESCRIPTION
**Pull request by devigned**
**What type of PR is this?**
/kind design

**What this PR does / why we need it**:
The scope of work for implementing safe rolling upgrades for AzureMachinePool with MaxUnavailable and MaxSurge has grown to warrant a proposal. This proposal provides background, tradeoffs and the design strategy to explain why we want to break out AzureMachinePoolMachines from AzureMachinePools.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #1105 

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Add AzureMachinePoolMachine proposal which describes the design strategy for implementing rolling upgrades for AzureMachinePools
```

### CODE REVIEW

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+
+```go
+// AzureMachinePoolSpec defines the desired state of AzureMachinePool
+AzureMachinePoolSpec struct {
+
+    /*
+       Other fields omitted for brevity
+    */
+	
+    // MaxUnavailable is the max number of replicas which can be unavailable at any time. For example, if the
```
Comment (CecileRobertMichon) [inline docs/proposals/20210222-azure-machinepool-machine.md:167]: My immediate reaction when seeing these new types is "why put them in CAPZ? there is nothing Azure infra specific about it". I know there are nuances and adding these types directly in CAPI is not straight forward because it then implies different providers would need to honor these values by either a) passing them through to the infra properties if the Cloud/environment supports them or b) implement it themselves (what we're doing in CAPZ). 

I worry that adding these here now will make it hard to move them later since any refactor of the types to CAPI types would be either a breaking change. And the alternative of leaving them in here beyond exp/ is that we create diverging implementations and, more importantly, inconsistent user experience for MachinePool across providers.

If we do go down this path, I suggest detailing this in a risks & mitigations section in this doc so that we can justify this design choice down the road when someone inevitably asks "why was this not added to CAPI types?"

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    //
+    // Note: Azure Subscriptions have a limited quota of CPUs. This quota can be expanded upon request. Due to the
+    // nature of this feature, spikes due to surge may cause you to exceed quota.
+    //
+    // See Also: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/azure-subscription-service-limits
+    MaxSurge *int32 `json:"maxSurge,omitempty"`
+}
+
+// AzureMachinePoolMachineSpec defines the desired state of AzureMachinePoolMachine
+AzureMachinePoolMachineSpec struct {
```
Comment (CecileRobertMichon) [inline docs/proposals/20210222-azure-machinepool-machine.md:189]: How do we make it clear to the users that they are not responsible for creating the AzureMachinePoolMachines?

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+
+**AzureMachinePool Responsibilities:**
+- Create AzureMachinePoolMachine instances when a new VMSS instance is observed. The AzureMachinePoolMachine spec should
+  have the `ProviderID` field set with the observed resource ID. The AzureMachinePool should also be added to the
+  AzureMachinePoolMachine's OwnerReferences. 
+- Upgrade / replace selection of AzureMachinePoolMachine instances. When a change to the AzureMachinePool model occurs,
+  the MachinePoolScope will be responsible for coordinating the rollout of the updated model by selecting
+  AzureMachinePoolMachines to delete / replace with respect to MaxUnavailable.
+- Scale up. AzureMachinePool should increase the number of VMSS replicas if the replica count increases on MachinePool
+- Scale down. AzureMachinePool should select and delete AzureMachinePoolMachines that are overprovisioned with respect
```
Comment (CecileRobertMichon) [inline docs/proposals/20210222-azure-machinepool-machine.md:277]: > should select

how does selection happen? at random? oldest first? newest first?

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+  - MaxUnavailable is the max number of machines that are allowed to be unavailable at any time
+  - MaxSurge is the number of machines to surge, add to the current replica count, during an upgrade to 
+- Safely update by cordoning and draining nodes prior to deleting the underlying infrastructure
+
+### Non-Goals / Future Work
+- Create a CAPI Machine owner for each AzureMachinePoolMachine (possible future work)
+
+## Proposal
+
+### User Stories
```
Comment (CecileRobertMichon) [inline docs/proposals/20210222-azure-machinepool-machine.md:76]: I'd love to see a user story for deleting a specific instance in a scale set

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+- Create AzureMachinePoolMachine instances when a new VMSS instance is observed. The AzureMachinePoolMachine spec should
+  have the `ProviderID` field set with the observed resource ID. The AzureMachinePool should also be added to the
+  AzureMachinePoolMachine's OwnerReferences. 
+- Upgrade / replace selection of AzureMachinePoolMachine instances. When a change to the AzureMachinePool model occurs,
+  the MachinePoolScope will be responsible for coordinating the rollout of the updated model by selecting
+  AzureMachinePoolMachines to delete / replace with respect to MaxUnavailable.
+- Scale up. AzureMachinePool should increase the number of VMSS replicas if the replica count increases on MachinePool
+- Scale down. AzureMachinePool should select and delete AzureMachinePoolMachines that are overprovisioned with respect
+  to MaxUnavailable.
+- Clean up. When a AzureMachinePoolMachine is no longer in the list of instances in Azure, but a matching
```
Comment (CecileRobertMichon) [inline docs/proposals/20210222-azure-machinepool-machine.md:279]: This implies Azure is the source of truth and the goal state. But at the same time, we're saying you can trigger a VMSS VM delete by deleting an AzureMachinePoolMachine which kind of goes against that assumption. IMO the CRs should be the source of truth / goal state, and the controller's job should be to reconcile against that goal state and try to get the real world (Azure resources) to match. Now if something happens out of band to delete a VMSS instance, the controller should see that and then our options are either 1) recreate it (that's what we do for all resources excepts VMs right now) or, and this is probably the option we need to pick if we're going to follow the immutable machines principle, 2) mark the AzureMachinePoolMachine as "Failed" and have some external remediation mechanism (equivalent to a MachineHealthCheck) who deletes failed machines - which should cause a new instance to be created to replace it since the desired replica count will become > than the actual number of replicas.

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+  AzureMachinePoolMachine resource exists, delete the AzureMachinePoolMachine.
+  
+**AzureMachinePoolMachine Responsibilities:**
+- Update Azure Provisioning State: when creating a new VMSS instance, the AzureMachinePoolMachine controller will poll 
+  the Azure API until the instance reaches a terminal state.
+- Cordon and Drain: when deleting the AzureMachinePoolMachine resource, the AzureMachinePoolMachine controller is
+  responsible for ensuring workload is moved from the node prior to removing the underlying Azure infrastructure.
+- NodeRef: as a VMSS instance joins the cluster, the AzureMachinePoolMachine controller is responsible for ensuring
```
Comment (CecileRobertMichon) [inline docs/proposals/20210222-azure-machinepool-machine.md:287]: The other behavior we need to define is here is what happens when I, a user, go and delete an AzureMachinePoolMachine? Does it get replaced by a new one (similar to MachineDeployment Machine behavior)? Or does the replica count get decreased by one (similar to VMSS delete instance behavior).

**Review by CecileRobertMichon — COMMENTED** [review]
Thank you so much for taking the time to write this down. It helped me understand better what the goals are in #1105, and I'm sure it will help future contributors and users as well.

Overall this looks good to me but I challenged some of the assumptions / implied design choices that might not be obvious to others. My main concern is around the delete scenario and how the sync works between Azure resource deletions and k8s CR deletions.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 2 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:189]: That's a great question. Do you have any suggestions?

My first inclination is to add documentation.

There really isn't anything that stops someone from creating an AMPM. It won't hurt anything.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 1 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:167]: Great observation. I totally agree. It'd be great to remove these from the AMP spec and move them to MP.

With https://github.com/kubernetes-sigs/cluster-api/pull/4073 nearly merged, I'll open up a PR to add a `RolloutStrategy` and `DeleteStrategy` to `MachinePool`.

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+    */
+	
+    // MaxUnavailable is the max number of replicas which can be unavailable at any time. For example, if the
+    // machine pool has a replica count of 5 and a MaxUnavailable value of 2, a rolling upgrade will never exceed
+    // more than 2 machines unavailable at a time.
+    //
+    // Note: If the MaxUnavailable value is equal to the machine pool replica value, there is
+    // the possibility of application down time.
+    // +kubebuilder:default=1
+    MaxUnavailable int32 `json:"maxUnavailable,omitempty"`
```
[the remark this answers is not in the document]
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:174]: Yes, they should.

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+Azure MachinePool currently embeds the state of each of the machines in the MachinePool within the status of the Azure
+MachinePool. MachinePool machines should be their own resources to enable individual lifecycles.
+
+## Motivation
+
+By giving each AzureMachinePoolMachine an individual lifecycle, a user would be able to inform CAPZ of the specific
+instance to delete and then have the AzureMachinePoolMachine controller cordon and drain the node prior to deleting
+the underlying infrastructure.
+
+Azure Virtual Machine Scale Set Instances do not provide the specific machine events needed to alert Kubernetes when
```
Comment (fiunchinho) [inline docs/proposals/20210222-azure-machinepool-machine.md:58]: Can't we leverage the [Azure scheduled events](https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-terminate-notification)?

**Comment by nader-ziada** [comment]
Thank you @devigned for this doc, very helpful in understanding how all this works

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 5 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:279]: Imagine when we over-provision VMs. Azure will spin up multiple, wait for the first to successfully deploy, then delete the others that didn't win the provisioning race. In this case, the loser VMs will be deleted by Azure. We must clean up the AMPMs which may have been created while the provisioning race was running.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 8 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:59]: > Terminate notifications only on ‘delete’ operations – All delete operations (manual delete or Autoscale-initiated scale-in) will generate Terminate events if your scale set has scheduledEventsProfile enabled. Other operations such as reboot, reimage, redeploy, and stop/deallocate do not generate Terminate events. 

We could if we decide to only delete and replace machines, which will cause infra to deprovision and then provision again. Another way we could do this is to use `Reimage`, which does not deprovision the infra, but only updates the model. In `Reimage`, we would not get an event from Azure letting the machine know that it should cordon and drain.

On the other hand, if we are using spot instances, we will likely need to listen to machine events to know when a node must cordon and drain.

This thread covers the problem pretty well: #819. I do think we will need some on host component like https://github.com/awesomenix/drainsafe at some point to help with spot instances or VM migration.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 4 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:79]: Added.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 3 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:332]: Added to take into account the DeletePolicy.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 6 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md:342]: Right now, the behavior is that the replica count stays the same. When an VMSS instance is deleted, the next time the AMP reconciler runs, it sees Azure with 1 less capacity than the MachinePool replica count. It PATCHs the capacity of the VMSS to be the replica count of MachinePool, which increases the capacity by 1. 

I think that's not what we want though, right? I think we want the capacity to decrease in that case, right?

Comment (CecileRobertMichon) [reply docs/proposals/20210222-azure-machinepool-machine.md:342]: > I think that's not what we want though, right? I think we want the capacity to decrease in that case, right?

To me, it makes sense to keep the replica count the same for the use case in user story #3. It's also more consistent with the MachineDeployment behavior. My comment is mostly that we should be explicit about that design choice, whether it's that or decreasing capacity.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 5 above)
Comment (CecileRobertMichon) [reply docs/proposals/20210222-azure-machinepool-machine.md:334]: That makes sense

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 1 above)
Comment (CecileRobertMichon) [reply docs/proposals/20210222-azure-machinepool-machine.md]: How is that action plan tracked in the proposal?

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 8 above)
Comment (fiunchinho) [reply docs/proposals/20210222-azure-machinepool-machine.md:59]: Oh I see. Thanks for the clarification. Still, I'd reword it a bit because it seems VMSS do provide machine events when an instance is deleted.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 1 above)
Comment (fiunchinho) [reply docs/proposals/20210222-azure-machinepool-machine.md]: It looks like [`AWSMachinePool` in CAPA contains similar fields](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/17c3144c5f0060526a005eb5a2927db22fcd259a/exp/api/v1alpha3/awsmachinepool_types.go#L82-L99). Not sure if it will be possible to move them to `MachinePool` as different cloud providers use very different mechanisms (ASG vs VMSS).

Comment (CecileRobertMichon) [reply docs/proposals/20210222-azure-machinepool-machine.md]: That's a good point @fiunchinho, it might make sense to keep them in cloud providers then. Let's document this as an intentional choice in this proposal.

**Comment by fiunchinho** [comment]
Looking at how [the upgrade process is implemented in KCP](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md#kubeadmcontrolplane-rollout) we could try to make the two look even more similar. There are two fields that I believe could be useful here:
- `nodeDrainTimeout` introduced in KCP [here](https://github.com/kubernetes-sigs/cluster-api/pull/3662), it allows the user to specify the amount of time we want to wait for a node to be drained. If capz is taking care of draining, I believe it should allow users to specify this.
- `upgradeAfter` introduced in KCP [here](https://github.com/kubernetes-sigs/cluster-api/pull/2238), is a field to indicate an upgrade should be performed after the specified time, to allow users to schedule upgrades.

What do you think?

**Comment by devigned** [comment]
> upgradeAfter introduced in KCP here, is a field to indicate an upgrade should be performed after the specified time, to allow users to schedule upgrades.

@fiunchinho I don't think that upgradeAfter makes sense for AzureMachinePool since if there are no changes to the underlying model, then there will be nothing to upgrade.

I added the `NodeDrainTimeout`. Thank you!

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 8 above)
Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md]: @fiunchinho wdyt of the termination notifications wording now?

**Comment by fiunchinho** [comment]
> > upgradeAfter introduced in KCP here, is a field to indicate an upgrade should be performed after the specified time, to allow users to schedule upgrades.
> 
> @fiunchinho I don't think that upgradeAfter makes sense for AzureMachinePool since if there are no changes to the underlying model, then there will be nothing to upgrade.
> 
> I added the `NodeDrainTimeout`. Thank you!

Totally with you. I miss understood the `upgradeAfter` field. Thanks.

File: docs/proposals/20210222-azure-machinepool-machine.md (hunk 8 above)
Comment (fiunchinho) [reply docs/proposals/20210222-azure-machinepool-machine.md]: Perfect.

File: docs/proposals/20210222-azure-machinepool-machine.md [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+    // The deployment strategy to use to replace existing machines with
+    // new ones.
+    // +optional
+    Strategy MachineDeploymentStrategy `json:"strategy,omitempty"`
+
+	// NodeDrainTimeout is the total amount of time that the controller will spend on draining a node.
+	// The default value is 0, meaning that the node can be drained without any time limitations.
+	// NOTE: NodeDrainTimeout is different from `kubectl drain --timeout`
+	// +optional
+	NodeDrainTimeout *metav1.Duration `json:"nodeDrainTimeout,omitempty"`
```
Comment (CecileRobertMichon) [inline docs/proposals/20210222-azure-machinepool-machine.md:264]: nit: indent

Comment (devigned) [reply docs/proposals/20210222-azure-machinepool-machine.md]: fixed

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
